File: server/static/js/pages/notify.js
// 主动触达页（推送通道 / 订阅 / 测试发送 / 投递记录）
// 依赖 core.js 的 state/api/toast/render/esc/fmt
//
// 演示要点：企业微信群机器人**不需要正式 AppID**，贴一个 webhook key 就能真的
// 收到消息；没配置时用本地记录通道（mock），消息会记到 data/notifications.jsonl，
// 页面下方「本地收件箱」能直接看到"推送了什么"。
'use strict';

async function loadNotify() {
  const n = state.notify;
  n.loading = true;
  render();
  const q = (p) => api(p).catch(() => null);
  const [prov, subs, logs, inbox] = await Promise.all([
    q('/api/notify/providers'), q('/api/notify/subscriptions'),
    q('/api/notify/logs'), q('/api/notify/mock-inbox'),
  ]);
  n.providers = (prov && prov.providers) || [];
  n.subs = (subs && subs.subscriptions) || [];
  n.logs = (logs && logs.logs) || [];
  n.inbox = (inbox && inbox.messages) || [];
  if (!n.form.channel && n.providers.length) n.form.channel = n.providers[0].id;
  n.loading = false;
  render();
}

function notifyPickChannel(id) {
  state.notify.form.channel = id;
  render();
}

async function saveSubscription() {
  const f = state.notify.form;
  if (!f.channel) { toast('先选一个通道'); return; }
  const needTarget = ['wecom_bot', 'wecom_app', 'wechat_subscribe'].includes(f.channel);
  if (needTarget && !f.target.trim()) {
    toast('这个通道需要填接收目标（如 webhook key）');
    return;
  }
  try {
    await api('/api/notify/subscriptions', 'POST', {
      channel: f.channel, target: f.target.trim(),
      name: f.name.trim(), events: f.events, enabled: true,
    });
    toast('已保存订阅');
    await loadNotify();
  } catch (e) { toast(e.message); }
}

async function testPush() {
  const f = state.notify.form;
  if (!f.channel) { toast('先选一个通道'); return; }
  try {
    const r = await api('/api/notify/test', 'POST', {
      channel: f.channel, target: f.target.trim(),
      title: '测试推送', content: '这是一条来自掌柜的测试消息',
    });
    const res = r.result || {};
    toast(res.ok ? '测试消息已发出' : ('发送失败：' + (res.error || '未知原因')));
    await loadNotify();
  } catch (e) { toast(e.message); }
}

async function delSubscription(sid) {
  if (!confirm('删除这条订阅？')) return;
  try { await api(`/api/notify/subscriptions/${sid}`, 'DELETE'); await loadNotify(); }
  catch (e) { toast(e.message); }
}

async function testSubscription(sid) {
  try {
    const r = await api(`/api/notify/subscriptions/${sid}/test`, 'POST', {});
    const res = r.result || {};
    toast(res.ok ? '测试消息已发出' : ('发送失败：' + (res.error || '未知原因')));
    await loadNotify();
  } catch (e) { toast(e.message); }
}

async function setupWecomBot() {
  const key = prompt('粘贴企业微信群机器人的 Webhook 地址（或只贴 key）\n\n微信：群设置 → 群机器人 → 添加 → 复制 Webhook 地址');
  if (!key) return;
  try {
    const r = await api('/api/notify/wecom-bot', 'POST', { key: key.trim() });
    toast(r.message || '已配置');
    await loadNotify();
  } catch (e) { toast(e.message); }
}

function notifyToggleEvent(ev) {
  const f = state.notify.form;
  const i = f.events.indexOf(ev);
  if (i >= 0) f.events.splice(i, 1); else f.events.push(ev);
  render();
}

// ---------------- 渲染 ----------------
const NOTIFY_EVENT_LABEL = {
  daily_review: '每日经营复盘',
  customer_reminder: '熟客提醒',
  payment_received: '收款到账',
  revenue_warning: '流水低于保本线预警',
};

function renderNotify() {
  const n = state.notify;
  const f = n.form;

  const chCards = n.providers.map(p => `
    <div class="pay-type-btn ${f.channel === p.id ? 'on' : ''}"
         style="text-align:left;padding:10px 12px;margin-bottom:6px"
         onclick="notifyPickChannel('${esc(p.id)}')">
      ${esc(p.name)} ${p.ready ? '<span class="pay-status">可用</span>' : '<span class="pay-status">待配置</span>'}
      <div class="pay-mchid">${esc(p.desc || p.requirement || '')}</div>
    </div>`).join('');

  const evBoxes = Object.keys(NOTIFY_EVENT_LABEL).map(ev => `
    <label class="acct-note" style="display:block;cursor:pointer">
      <input type="checkbox" ${f.events.includes(ev) ? 'checked' : ''}
             onchange="notifyToggleEvent('${ev}')" />
      ${NOTIFY_EVENT_LABEL[ev]}</label>`).join('');

  const subRows = n.subs.length === 0
    ? '<div class="empty">还没有订阅</div>'
    : n.subs.map(s => `
      <div class="pay-row">
        <div class="pay-row-main">
          <div class="pay-type-badge">${esc(s.channel)}</div>
          <div class="txn-sub">${esc(s.target || '默认目标')}</div>
          <div class="pay-mchid">${(s.events || []).map(e => NOTIFY_EVENT_LABEL[e] || e).join('、') || '未选事件'}</div>
        </div>
        <div class="pay-row-actions">
          <button class="btn-mini" onclick="testSubscription(${s.id})">测试</button>
          <button class="btn-mini btn-danger" onclick="delSubscription(${s.id})">删除</button>
        </div>
      </div>`).join('');

  const inbox = n.inbox.length === 0
    ? '<div class="empty">本地通道还没收到消息，点上面「发一条测试推送」试试</div>'
    : n.inbox.slice(0, 8).map(m => `
      <div class="review-box" style="margin-bottom:8px">
        <strong>${esc(m.title || '')}</strong><br/>${esc(m.content || '')}
        <div class="txn-time">${esc(m.created_at || '')}</div>
      </div>`).join('');

  const logs = n.logs.length === 0
    ? '<div class="empty">暂无投递记录</div>'
    : n.logs.slice(0, 12).map(l => `
      <div class="txn-row">
        <div class="txn-main">
          <div class="txn-item">${esc(NOTIFY_EVENT_LABEL[l.event] || l.event)}
            ${l.ok ? '<span class="pay-status">成功</span>' : '<span class="txn-amount expense">失败</span>'}</div>
          <div class="txn-sub">${esc(l.channel || '')} · ${esc(l.created_at || '')}${l.retries ? ' · 重试 ' + l.retries + ' 次' : ''}</div>
          ${l.error ? `<div class="pay-mchid">${esc(l.error)}</div>` : ''}
        </div>
      </div>`).join('');

  return `
  <div class="hero"><div class="hero-title">主动触达</div>
    <div class="hero-sub">复盘与提醒主动推到手机上，不用自己想起来打开</div></div>

  <div class="card">
    <div class="card-title">⚡ 最快接通：企业微信群机器人</div>
    <div class="acct-note">不需要正式小程序 AppID。手机企业微信建一个只有自己的群，
      群设置 → 群机器人 → 添加 → 复制 Webhook 里的 key 即可。</div>
    <button class="btn-primary" onclick="setupWecomBot()">一步配置并立即测试</button>
  </div>

  <div class="card">
    <div class="card-title">📡 推送通道</div>
    ${chCards || '<div class="empty">没有可用通道</div>'}
    <div class="form-item"><label class="form-label">接收目标（webhook key / 用户 ID）</label>
      <input class="form-input" type="text" placeholder="本地记录通道可留空"
             value="${esc(f.target)}" oninput="state.notify.form.target=this.value" /></div>
    <div class="form-item"><label class="form-label">订阅哪些消息</label>${evBoxes}</div>
    <div class="pay-actions">
      <button class="btn-mini" onclick="testPush()">发一条测试推送</button>
      <button class="btn-mini" onclick="saveSubscription()">保存订阅</button>
    </div>
  </div>

  <div class="card">
    <div class="card-title">🔔 已订阅（${n.subs.length}）</div>
    ${subRows}
  </div>

  <div class="card">
    <div class="card-title">📥 本地收件箱（演示用）</div>
    ${inbox}
  </div>

  <div class="card">
    <div class="card-title">📜 投递记录</div>
    ${logs}
  </div>`;
}

File: server/routers/notify.py
"""主动触达：消息订阅 / 推送测试 / 投递记录

对应缺口：原先所有能力都是被动的 —— 店主必须自己想起来打开小程序才能看到
每日复盘与熟客提醒。本路由让这些内容能主动推到店主的手机上。

通道选择建议（见 GET /api/notify/providers）：
  - wecom_bot：企业微信群机器人，**不需要正式 AppID**，配个 webhook 就能收到真消息
  - mock：本地记录，无需任何配置，用于演示与自测
  - wecom_app / wechat_subscribe：需要企业微信应用或正式小程序 AppID
"""
import logging

from fastapi import APIRouter, HTTPException, Query

import notifications
from schemas import NotifySubscriptionIn, NotifyTestIn, WecomBotIn

log = logging.getLogger("routers.notify")
router = APIRouter(prefix="/api", tags=["notify"])


@router.get("/notify/events")
def notify_events():
    """可订阅的事件类型（前端拿来渲染勾选框）。"""
    return {"events": notifications.list_events()}


@router.get("/notify/providers")
def notify_providers():
    """可用通道及其配置要求 —— 让用户知道哪条能立刻用、哪条要资质。"""
    return {"providers": [
        {"id": "mock", "name": notifications.PROVIDER_NAMES["mock"],
         "ready": True, "target_label": "", "hint": "无需配置，推送内容记录在本地文件"},
        {"id": "webhook", "name": notifications.PROVIDER_NAMES["webhook"],
         "ready": True, "target_label": "https://your-host/path 或 ...|共享密钥",
         "hint": "开放 API：把事件以 JSON POST 出去；填 |密钥 则附 HMAC-SHA256 签名，供接收方验签"},
        {"id": "wecom_bot", "name": notifications.PROVIDER_NAMES["wecom_bot"],
         "ready": True, "target_label": "群机器人 Webhook key",
         "hint": "企业微信群 → 添加群机器人 → 复制 Webhook 地址里的 key；不需要正式 AppID"},
        {"id": "wecom_app", "name": notifications.PROVIDER_NAMES["wecom_app"],
         "ready": False, "target_label": "corpid:secret:agentid:用户",
         "hint": "需企业微信管理员创建应用"},
        {"id": "wechat_subscribe", "name": notifications.PROVIDER_NAMES["wechat_subscribe"],
         "ready": False, "target_label": "appid:secret:openid:template_id",
         "hint": "需正式 AppID、用户逐次授权订阅、并在公众平台申请模板；游客模式无法使用"},
    ]}


@router.get("/notify/subscriptions")
def notify_subscriptions():
    return {"subscriptions": notifications.list_subscriptions()}


@router.post("/notify/subscriptions")
def notify_subscription_save(data: NotifySubscriptionIn):
    """新增/更新订阅（通道 + 目标 + 订阅哪些事件）。"""
    try:
        sub = notifications.save_subscription(
            data.channel, data.target, data.events or [],
            enabled=data.enabled, name=data.name, sid=data.sid)
    except ValueError as e:
        raise HTTPException(400, str(e)) from e
    return {"ok": True, "subscription": sub}


@router.delete("/notify/subscriptions/{sid}")
def notify_subscription_delete(sid: int):
    if not notifications.delete_subscription(sid):
        raise HTTPException(404, "订阅不存在")
    return {"ok": True}


@router.post("/notify/subscriptions/{sid}/test")
def notify_subscription_test(sid: int):
    """给某个订阅发一条测试消息，立刻验证通道是否通。"""
    subs = [s for s in notifications.list_subscriptions() if s["id"] == sid]
    if not subs:
        raise HTTPException(404, "订阅不存在")
    s = subs[0]
    r = notifications.notify(
        s["channel"], s["target"], "【测试】巷子里的AI掌柜",
        "这是一条测试消息。收到说明推送通道配置正确。", event="manual")
    if not r["ok"]:
        raise HTTPException(502, f"推送失败：{r.get('error')}")
    return {"ok": True, "result": r}


@router.post("/notify/test")
def notify_test(data: NotifyTestIn):
    """不建订阅，直接试发一条（用于快速验证通道）。"""
    r = notifications.notify(data.channel, data.target,
                             data.title or "【测试】巷子里的AI掌柜",
                             data.content or "推送通道测试成功。", event="manual")
    if not r["ok"]:
        raise HTTPException(502, f"推送失败：{r.get('error')}")
    return {"ok": True, "result": r}


@router.post("/notify/wecom-bot")
def notify_set_wecom_bot(data: WecomBotIn):
    """一步配置企业微信群机器人（最省事的真实通道）。

    自动给全部事件建好订阅，并立即发一条测试消息确认通道可用。
    """
    key = (data.key or "").strip()
    if not key:
        raise HTTPException(400, "请填写群机器人 Webhook key")
    try:
        sub = notifications.save_subscription(
            "wecom_bot", key, data.events or [],
            enabled=True, name=data.name or "企业微信群机器人", sid=data.sid)
    except ValueError as e:
        raise HTTPException(400, str(e)) from e
    test = notifications.notify(
        "wecom_bot", key, "【已接通】巷子里的AI掌柜",
        "推送通道已配置成功，之后每天的经营复盘与熟客提醒会发到这里。",
        event="manual")
    return {"ok": True, "subscription": sub, "test": test,
            "message": "已接通" if test["ok"] else f"订阅已保存，但测试消息发送失败：{test.get('error')}"}


@router.get("/notify/logs")
def notify_logs(limit: int = Query(default=50, ge=1, le=500), event: str | None = None):
    """投递记录（成功/失败/重试次数/错误都能查，便于排查通道问题）。"""
    try:
        return {"logs": notifications.list_logs(limit=limit, event=event)}
    except ValueError as e:
        raise HTTPException(400, str(e)) from e


@router.get("/notify/mock-inbox")
def notify_mock_inbox(limit: int = Query(default=20, ge=1, le=200)):
    """本地记录通道收到的消息（演示时用它直观展示"推送到了什么"）。"""
    return {"messages": notifications.recent_mock_messages(limit=limit)}


@router.post("/notify/dispatch")
def notify_dispatch(event: str = Query(...), title: str = Query(default=""),
                    content: str = Query(default=""),
                    business_key: str = Query(default=""),
                    dedup_days: int = Query(default=0, ge=0, le=30)):
    """手动触发一次事件分发（调试与演示用）。

    business_key + dedup_days 可做幂等去重：同一业务键在 N 天内只推一次。
    定时任务用的就是这套（如按日期去重，避免服务重启后重复推送刷屏）。
    """
    try:
        return notifications.dispatch_event(
            event, title or "【手动触发】", content or "",
            business_key=business_key, dedup_days=dedup_days)
    except ValueError as e:
        raise HTTPException(400, str(e)) from e
